app/javascript/lib/scatter-plot.js

```diff
@@ -0,0 +1,308 @@
+    if (typeof cameraData['scene.camera'] !== 'undefined') {
+      const newCamera = cameraData['scene.camera']
+      $(`#${plotId}`).data('camera', newCamera)
+    }
+  })
+}
+
+/**
+ * Load and draw scatter plot.  Handles clusters and spatial groups, numeric
+ * and group annotations.
+ *
+ * @param {Object} clusterParams Parameters for `/clusters` plot API endpoint
+ * @param {Object} frame Plot UI properties not from plot API data
+ */
+export async function scatterPlot(clusterParams, frame) {
+  // Copy by value; preserves properties (e.g. plotId) across `await`
+  frame = Object.assign({}, frame)
+
+  const { plotId, hasLegend } = frame
+  console.log(plotId)
+
+  const legendId = (hasLegend ? `${plotId}-legend` : null)
+  const legendHtml = (hasLegend ? `<div id="${legendId}"></div>` : '')
+
+  $(frame.selector).append(
+    `<div id="${plotId}" class="plot"></div>
+      ${legendHtml}
+    </div>`
+  )
+
+  // Set default number of rows and columns of plots
+  if (!frame.numRows) {frame.numRows = 1}
+  if (!frame.numColumns) {frame.numColumns = 1}
+
+  const $plotElement = $(`#${plotId}`)
+  const spinnerTarget = $plotElement[0]
+  const spinner = new Spinner(window.opts).spin(spinnerTarget)
+  // $plotElement.data('spinner', spinner)
+
+  const { accession, cluster, annotation, subsample, gene } = clusterParams
+  let isAnnotatedScatter = null
+  if (
+    'isAnnotatedScatter' in clusterParams && clusterParams.isAnnotatedScatter
+  ) {
+    isAnnotatedScatter = true
+  }
+
+  $('#search_annotation').val(annotation)
+  $('#gene_set_annotation').val(annotation)
+
+  let rawPlot = await fetchCluster(
+    accession, cluster, annotation, subsample, null, gene, isAnnotatedScatter
+  )
+  rawPlot = Object.assign(rawPlot, frame)
+
+  // Consider putting into a dictionary instead of a list
+  window.SCP.scatterPlots.push(rawPlot)
+
+  // render annotation toggler picker if needed
+  if (rawPlot.annotParams.type == 'numeric') {
+    $('#toggle-plots').html('')
+  } else {
+    // Consider restoring this; it was missing on production before refactor
+    // $('#toggle-plots').html(
+    //   '<a href="#" class="btn btn-default btn-sm" id="toggle-traces" ' +
+    //     'data-toggle="tooltip" data-placement="left" data-trigger="hover" ' +
+    //     'title="Click to toggle all annotations, or toggle individual ' +
+    //       'annotations by clicking the legend entry"' +
+    //   '>Toggle Annotations <i class="fas fa-toggle-on"></i></a>'
+    // )
+    // $('#toggle-traces').tooltip({
+    //   container: 'body', placement: 'left', trigger: 'hover'
+    // })
+  }
+
+  renderScatterPlot(rawPlot, plotId, legendId)
+
+  spinner.stop()
+}
+
+/**
+ * Load and draw scatter plots for reference clusters, gene + ref clusters
+ *
+ * @param {Object} study Study object returned by clusters API endpoint
+ * @param {String} gene Searched gene name, e.g. "TP53"
+ * @param {Boolean} hasReference Whether each plot has a paired reference
+ */
```

Comment:
This function seems to mix the business logic of deciding which plots to render for a given study/gene combo with the rendering logic of 'how to show plots in a grid'.  And I'm not sure either of those belongs in this file, since both may be somewhat page-specific.  For now, I'd just move this function to one of the explore-* js files, so that it's clearer that this function has opinions about what plots it will draw.

Reply:
Good point.  This function was a vestige from earlier refactoring.  I moved it to a new module, `explore.js`.